syft/federated/federated_client.py

```diff
@@ -0,0 +1,136 @@
+import torch as th
+from torch import nn
+from torch.utils.data import BatchSampler, RandomSampler, SequentialSampler
+import logging
+
+from syft.generic import ObjectStorage
+from syft.federated.train_config import TrainConfig
+
+logger = logging.getLogger(__name__)
+
+
+class FederatedClient(ObjectStorage):
+    """A Client able to execute federated learning in a defined dataset."""
+
+    def __init__(self, datasets=None, verbose=True):
+        super().__init__()
+        self.datasets = datasets if datasets is not None else dict()
+        self.optimizer = None
+        self.train_config = None
+        self.parameters = []
+        self.verbose = verbose
+
+    def add_dataset(self, dataset, key: str):
+        self.datasets[key] = dataset
+
+    def remove_dataset(self, key: str):
+        if key in self.datasets:
+            del self.datasets[key]
+
+    def set_obj(self, obj: object):
+        """Registers objects checking if which objects it should cache.
+
+        Args:
+            obj: An object to be registered.
+        """
+        if isinstance(obj, TrainConfig):
+            self.train_config = obj
+            self.optimizer = None
+        elif isinstance(obj, nn.parameter.Parameter):
+            self.parameters.append(obj)
+
+        super().set_obj(obj)
+
+    def _build_optimizer(self, optimizer_name: str, model, lr: float) -> th.optim.Optimizer:
+        """Build an optimizer if needed.
+
+        Args:
+            optimizer_name: A string indicating the optimizer name.
+            lr: A float indicating the learning rate.
+        Returns:
+            A Torch Optimizer.
+        """
+        if self.optimizer is not None:
+            return self.optimizer
+
+        optimizer_name = optimizer_name.lower()
+        if optimizer_name == "sgd":
+            logger.debug("sgd parameters = %s, lr = %s", model.parameters(), lr)
+            self.optimizer = th.optim.SGD(model.parameters(), lr=lr)
+        else:
+            raise ValueError("Unknown optimizer: {}".format(optimizer_name))
+        return self.optimizer
+
+    def fit_batch(self, data, target, *args, **kwargs):
+        if self.verbose:
+            logger.debug("data = %s", data)
+            logger.debug("target = %s", target)
+        self.register_obj(data)
```

Comment:
I don't like that we need to register/deregister tensors explicitly.

Reply:
The fit_batch method still relies on the previous train_config content with forward_plans etc.. I will update it. The call to register the data will go away. It was needed for plans to work.